src/librustc_infer/traits/wf.rs

```diff
@@ -604,11 +603,14 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
         substs: SubstsRef<'tcx>,
     ) -> Vec<traits::PredicateObligation<'tcx>> {
         let predicates = self.infcx.tcx.predicates_of(def_id).instantiate(self.infcx.tcx, substs);
-        let cause = self.cause(traits::ItemObligation(def_id));
         predicates
             .predicates
             .into_iter()
-            .map(|pred| traits::Obligation::new(cause.clone(), self.param_env, pred))
+            .zip(predicates.spans.into_iter())
+            .map(|(pred, span)| {
+                let cause = self.cause(traits::BindingObligation(def_id, span));
```

Comment:
Before I forget, cc @nikomatsakis on the `ItemObligation` vs `BindingObligation` split, and the name of the latter.

Reply:
https://github.com/rust-lang/rust/pull/69709#discussion_r388460760

Reply:
Hmm, the name `BindingObligation` doesn't say that much to me, and the [rustdoc on it](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/traits/enum.ObligationCauseCode.html#variant.BindingObligation) is not very precise:

> Like ItemObligation, but with extra detail on the source of the obligation.

I presume that it's meant to track the precise bound that caused this obligation? In which case I think the name is "decent", but I'd probably just call it `WhereClause` or something. I don't think the suffix "Obligation" is adding much here.